### ISSUE TITLE
Efficient softmax implementation with Eigen::EigenMatrixXf<float, RowMajor>

### ISSUE THREAD

**Issue by tmr-monolith** [tmrob2/low-level-deep-learning#3]
For the cross entropy loss function, an efficient implementation of the softmax of a matrix is required.

Proposed implementation: 

Assuming that the input is a dynamic Eigen matrix of floating point type in `RowMajor` order 
```c++
auto wMinusMax = (input.rowwise() - input.colwise().maxCoeff()).array() // to control numerical instability
wMinusMax.rowwise() - wMinusMax.exp().colwise().sum().log() // this is using a log trick to reorder the computation
```

## Testing:

Create an interface to the softmax using a submodule so that we don't expose too many functions to the module:

```c++
py::module_ loss_functions = m.def_submodule("loss_functions", "some text explaining the submodule")

loss_functions.def("softmax", &nn2::loss::softmax);
```

Create a Python test

```Python
import cppapi
import numpy as np
def test_softmax():
    x = np.random.randn(5, 3).astype(np.float32)
    softmax_np = np.exp(x)/sum(np.exp(x))
    softmax_cpp = cppapi.loss_functions.softmax(x)
    assert(np.allclose(softmax_np, softmax_cpp))
```
